applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.h

```diff
@@ -0,0 +1,524 @@
+
+    ///@}
+    ///@name Access
+    ///@{
+
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const override
+    {
+        return "MapperVertexMorphingSymmetric";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "MapperVertexMorphingSymmetric";
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const override
+    {
+    }
+
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    // Initialized by class constructor
+    ModelPart& mrOriginModelPart;
+    ModelPart& mrDestinationModelPart;
+    Parameters mMapperSettings;
+    FilterFunction::Pointer mpFilterFunction;
+    bool mIsMappingInitialized = false;
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    virtual void InitializeComputationOfMappingMatrix()
+    {
+        mpSearchTree.reset();
+        mMappingMatrix.clear();
+    }
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+
+    ///@}
+
+private:
+    ///@name Static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // Variables for spatial search
+    unsigned int mBucketSize = 100;
+    KDTree::Pointer mpSearchTree;
+
+    // Variables for mapping
+    SparseMatrixType mMappingMatrix;
+
+    SymmetryBase::Pointer mpSymmetry;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    // --------------------------------------------------------------------------
+    void CreateFilterFunction()
+    {
+        const std::string filter_type = mMapperSettings["filter_function_type"].GetString();
+        const double filter_radius = mMapperSettings["filter_radius"].GetDouble();
+
+        mpFilterFunction = Kratos::shared_ptr<FilterFunction>(new FilterFunction(filter_type, filter_radius));
+    }
+
+    // --------------------------------------------------------------------------
+    void InitializeMappingVariables()
+    {
+        const unsigned int origin_node_number = mrOriginModelPart.Nodes().size();
+        const unsigned int destination_node_number = mrDestinationModelPart.Nodes().size();
+        mMappingMatrix.resize(destination_node_number*3,origin_node_number*3,false);
+    }
+
+    // --------------------------------------------------------------------------
+    void AssignMappingIds()
+    {
+        unsigned int i = 0;
+        for(auto& node_i : mrOriginModelPart.Nodes())
+            node_i.SetValue(MAPPING_ID,i++);
+
+        i = 0;
+        for(auto& node_i : mrDestinationModelPart.Nodes())
+            node_i.SetValue(MAPPING_ID,i++);
+    }
```

Comment:
So `MAPPING_ID` is unique for each node ryt?

Reply:
yes